library/core/src/ptr/non_null.rs

```diff
@@ -502,6 +502,11 @@ impl<T: ?Sized> NonNull<T> {
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
     #[stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_const_stable(feature = "non_null_convenience", since = "1.80.0")]
+    #[requires(!origin.as_ptr().is_null())]
+    #[ensures(
```

Comment:
Consider checking that the two pointers are still pointing to the same object(when this approach becomes available)

Reply:
TODO: move it in contract after using latest kani built from source code